core: remount sysfs when network and mount namespace are enabled

yuwata opened the pull request:
Fixes #26422.

poettering commented on src/shared/mount-util.c:
```diff
@@ -1167,6 +1168,161 @@ int remount_idmap(
+        size_t n;
+        int r;
+
+        assert(path);
```
Maybe add a quick comment what this actually does, i.e. why we need something so involved here. It's not obvious to the random reader I think.

poettering commented on man/systemd.exec.xml:
```diff
@@ -1663,6 +1663,10 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
         security.</para>
 
+        <para>When this option is enabled, <varname>PrivateMounts=</varname> is also enabled unless it is
+        explicitly disabled, and <filename>/sys</filename> will be remounted to make it tagged with the new
```
"make it tagged with" → "associate it with"?

poettering commented on man/systemd.exec.xml:
```diff
@@ -1684,6 +1688,10 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
 
+        <para>When this option is enabled, <varname>PrivateMounts=</varname> is also enabled unless it is
+        explicitly disabled, and <filename>/sys</filename> will be remounted to make it tagged with the new
+        network namespace.</para>
```
ditto

poettering commented on man/systemd.exec.xml:
```diff
@@ -1684,6 +1688,10 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
         <varname>NetworkNamespacePath=</varname> configured, as otherwise the network namespace of those
         units is reused.</para>
 
+        <para>When this option is enabled, <varname>PrivateMounts=</varname> is also enabled unless it is
```
"also enabled" → "implied"?

(similare above)

bluca commented:
```
../src/shared/mount-util.c: In function ‘remount_sysfs’:
../src/shared/mount-util.c:1286:30: error: ‘n’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
 1286 |         for (size_t i = 0; i < n; i++) {
      |                            ~~^~~
../src/shared/mount-util.c:1302:16: note: ‘n’ was declared here
 1302 |         size_t n;
      |                ^
../src/shared/mount-util.c:1301:19: error: ‘mounts’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
 1301 |         SubMount *mounts;

```

yuwata commented:
@poettering and @bluca Thank you for your comments. All suggestions are addressed. Moreover,
- also introduced `exec_needs_ipc_namespace()` helper function,
- fixed missing condition for creating ExecRuntime object,
- made `remount_sysfs()` more generic, 
- and more tests.

PTAL.

poettering commented on src/shared/mount-util.c:
```diff
@@ -1167,6 +1168,182 @@ int remount_idmap(
+                struct libmnt_fs *fs;
+                const char *path;
+
+                r = mnt_table_next_fs(table, iter, &fs);
```
hmm, i guess we could use mnt_table_next_child_fs() here?

yuwata replied:
Unfortunately, it cannot. as child mount may be overmounted.

poettering commented on src/shared/mount-util.c:
```diff
@@ -1167,6 +1168,182 @@ int remount_idmap(
+
+fail:
+        assert(r < 0);
+        sub_mount_free_n(mounts, n);
```
candidate for CLEAN_ARRAY() → #26555

yuwata replied:
Nice. I will apply it after the PR is merged.

dechamps commented:
@yuwata It would have been nice to have this be mentioned in [`NEWS`](https://github.com/systemd/systemd/blob/main/NEWS), as `PrivateNetwork=yes` implying `PrivateMounts=yes` (see c2da3bf2376440526e87da6124660d22ac2079d6) is a non-trivial change in behaviour that [caught some people off-guard](https://github.com/systemd/systemd/issues/2741#issuecomment-1661048428) :(
